Fixes #3894: Unable to exit overflow menu (exploration + revision card)

vrajdesai78 opened the pull request:
Fixes #3894: Replaced overflow menu with bottomsheet so that when talkback is turned on user can able to exit from it by pressing close. 

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### Passing test cases for ExplorationActivity

![Screenshot from 2022-08-02 00-04-31](https://user-images.githubusercontent.com/43074241/182221169-0cb1e5e9-507f-41b8-8a66-83262c95ad7c.png)
![Screenshot from 2022-08-02 00-05-36](https://user-images.githubusercontent.com/43074241/182221184-9dc3e77d-c2ae-4098-9d31-aa086baebc30.png)
![Screenshot from 2022-08-02 00-06-09](https://user-images.githubusercontent.com/43074241/182221193-34d50a90-bb24-4b21-b1d4-95d01128ef09.png)
![Screenshot from 2022-08-02 00-12-56](https://user-images.githubusercontent.com/43074241/182221222-10128252-f545-4a3f-bee5-f76106c574dd.png)

### Passing test cases for RevisionCardActivity
![Screenshot from 2022-08-18 23-56-03](https://user-images.githubusercontent.com/43074241/185467761-f49ac7c3-d92e-4965-8cee-9986777d2b86.png)

### RTL Screenshots (Screenshot taken after adding multilangauge string for close in bottomsheet)

![R](https://user-images.githubusercontent.com/43074241/185666732-ca8f2784-b246-4e9d-9ad3-1138ebbd4103.jpeg)
![RTL1](https://user-images.githubusercontent.com/43074241/185666768-6c1ce2f1-3c3b-4fd6-a398-2c5df602ac2f.jpeg)


### Video


https://user-images.githubusercontent.com/43074241/185469039-24e440da-69f9-4885-938c-688e3ff3ecf6.mp4

vrajdesai78 committed:
bottomsheet added

vrajdesai78 committed:
Added test cases and optimized code

vrajdesai78 committed:
Nit changes

vrajdesai78 committed:
updated id

vrajdesai78 committed:
Merge branch 'develop' into Bottomsheet-for-overflow-menu

vrajdesai78 committed:
klint issue fixed

vrajdesai78 committed:
Fixed failing static check

vrajdesai78 committed:
Bazel syntax fixed

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/BottomSheetOptionsMenu.kt:
```diff
@@ -0,0 +1,38 @@
+import org.oppia.android.R
+import org.oppia.android.databinding.BottomSheetOptionsMenuFragmentBinding
+
+/** Bottom sheet fragment for displaying options menu */
```
Nit: `/** Bottom sheet fragment for displaying options menu. */

vrajdesai78 replied:
Done

rt4914 commented on app/src/main/res/drawable/ic_bottomsheet.xml:
```diff
@@ -0,0 +1,5 @@
+  android:viewportHeight="24" android:viewportWidth="24"
+  android:width="24dp" xmlns:android="http://schemas.android.com/apk/res/android">
+  <path android:fillColor="#fff" android:pathData="M12,8c1.1,0 2,-0.9 2,-2s-0.9,-2 -2,-2 -2,0.9 -2,2 0.9,2 2,2zM12,10c-1.1,0 -2,0.9 -2,2s0.9,2 2,2 2,-0.9 2,-2 -0.9,-2 -2,-2zM12,16c-1.1,0 -2,0.9 -2,2s0.9,2 2,2 2,-0.9 2,-2 -0.9,-2 -2,-2z"/>
+</vector>
```
EOF is missing, add one blank line at the end.

Press `Control + Alt + L` - reformatting shortcut to fix the formatting.

vrajdesai78 replied:
Updated

rt4914 commented on app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml:
```diff
@@ -0,0 +1,44 @@
+      android:textSize="20sp"
+      app:drawableStartCompat="@drawable/ic_baseline_close_24" />
+  </androidx.appcompat.widget.LinearLayoutCompat>
+</layout>
```
EOF is missing, add one blank line at the end.

vrajdesai78 replied:
Updated

rt4914 commented on app/src/main/res/layout/exploration_activity.xml:
```diff
@@ -55,6 +54,15 @@
+            <ImageButton
+              android:id="@+id/action_bottom_sheet_options_menu"
+              android:layout_width="48dp"
+              android:layout_height="48dp"
```
These should be `wrap_content` and you should use `minHeight=48dp` and `minWidth=48dp`

vrajdesai78 replied:
Updated

rt4914 commented on app/src/main/res/layout/exploration_activity.xml:
```diff
@@ -55,6 +54,15 @@
+              android:layout_height="48dp"
+              android:layout_gravity="end"
+              android:scaleType="center"
+              style="?android:actionOverflowButtonStyle"/>
```
I believe `style` should be the first attribute under `ImageButton`. 
Press `Control + Alt + L` - reformatting shortcut to fix the formatting.

rt4914 replied:
Also, `space` is missing before the closing tag.

vrajdesai78 replied:
Updated

rt4914 commented on app/src/main/res/values/strings.xml:
```diff
@@ -6,6 +6,7 @@
   <string name="menu_help">Help</string>
   <string name="exploration_activity_title">Exploration Player</string>
   <string name="help_activity_title">Help</string>
+  <string name="bottom_sheet_options_menu_close" translatable="false">Close</string>
```
Why `translatable = false`? In other languages it should be read as different.

vrajdesai78 replied:
Removed it

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/BottomSheetOptionsMenu.kt:
```diff
@@ -0,0 +1,38 @@
+import org.oppia.android.databinding.BottomSheetOptionsMenuFragmentBinding
+
+/** Bottom sheet fragment for displaying options menu */
+class BottomSheetOptionsMenu : BottomSheetDialogFragment() {
```
`InjectableDialogFragment ` should we have a `InjectableBottomSheetDialogFragment` too which should be used here. For reference check any dialog fragment.

vrajdesai78 replied:
Added InjectableBottomSheetDialogFragment

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/BottomSheetOptionsMenuItemClickListener.kt:
```diff
@@ -0,0 +1,14 @@
+
+/**
+ * Listener for when the user interacts with the play session options bottom sheet.
+ * [BottomSheetOptionsMenu].
```
This looks odd. After `.` in this previous line why is `[BottomSheetOptionsMenu]` mentioned. It should be one meaningful sentence.

vrajdesai78 replied:
Updated

rt4914 commented on app/src/main/java/org/oppia/android/app/player/exploration/BottomSheetOptionsMenuItemClickListener.kt:
```diff
@@ -0,0 +1,14 @@
+   *
+   * @param itemId resource Id of the option selected
+   */
+  fun handleOnOptionsItemSelected(itemID: Int)
```
Should be `itemId `

vrajdesai78 replied:
Updated

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -1713,6 +1713,111 @@ class ExplorationActivityTest {
+  }
+
+  @Test
+  fun testActivity_openBottomsheet_selectCloseInBottomSheet_closeBottomsheet() {
```
`testExplorationActivity_openBottomsheet_selectCloseOption_bottomSheetCloses`

vrajdesai78 replied:
Done

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -1713,6 +1713,111 @@ class ExplorationActivityTest {
   }
 
+  @Test
+  fun testActivity_initialise_openBottomSheet_showsBottomSheet() {
```
`testActivity` -> `testExplorationActivity`

vrajdesai78 replied:
Done

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -1713,6 +1713,111 @@ class ExplorationActivityTest {
+  }
+
+  @Test
+  fun testExploration_openBottomsheet_selectHelpInBottomsheet_opensHelpActivity() {
```
`testExplorationActivity_...`

vrajdesai78 replied:
Done

rt4914 reviewed (changes requested):
@vrajdesai78 
Please do one of these:
1. Either name this PR as `Fix part of #xxxx: Replace BottomSheet in place of Overflow menu in ExplorationActivity`. and in another PR work on Revision Card
2. Or Name it `Fix  #xxxx: Replace BottomSheet in place of Overflow menu in ExplorationActivity and RevisionActivity`. + Add revision card code, output etc. to this PR.

I would suggest go with (1.) so that the second PR can be done really easily.

vrajdesai78 committed:
InjectableBottomSheetDialogFragment added and done nit changes

vrajdesai78 committed:
Merge branch 'develop' into Bottomsheet-for-overflow-menu

vrajdesai78 committed:
Fixed failing CI checks

vrajdesai78 committed:
nit changes

vrajdesai78 commented:
I have made suggested changes by @rt4914. Still don't know why CI checks are failing. I have tried to run robolectric test of ExplorationActivityTest and they are passing.

rt4914 commented:
@vrajdesai78 I think in this either @BenHenning or someone with experience with Bazel will need to review it as I think most of the failing checks are related to bazel only.

rt4914 commented:
@vrajdesai78 Please check this https://github.com/oppia/oppia-android/pull/4482#pullrequestreview-1060335293 and also reply to all open comments.

vrajdesai78 committed:
Nit changes

vrajdesai78 commented:
> @vrajdesai78 Please do one of these:
> 
> 1. Either name this PR as `Fix part of #xxxx: Replace BottomSheet in place of Overflow menu in ExplorationActivity`. and in another PR work on Revision Card
> 2. Or Name it `Fix  #xxxx: Replace BottomSheet in place of Overflow menu in ExplorationActivity and RevisionActivity`. + Add revision card code, output etc. to this PR.
> 
> I would suggest go with (1.) so that the second PR can be done really easily.

I have updated PR title and will create a new PR for revision card once this PR get merged

yash10019coder commented on app/src/main/java/org/oppia/android/app/fragment/BUILD.bazel:
```diff
@@ -91,3 +91,18 @@ kt_android_library(
     ],
 )
+
+kt_android_library(
```
please include this target in [app module library in app/BUILD.bazel](https://github.com/yash10019coder/oppia-android/blob/d8ab9e953f694e21fb6621ef7786516f9f068bce/app/BUILD.bazel#L734)  as the dependency. It's required fragment for the app.

vrajdesai78 replied:
Added, Thanks

yash10019coder reviewed (changes requested):
Hi @vrajdesai78 most of the things look good to me in your PR it's just you need to include the InjectableBottomSheetDialog fragmnet as a dependency in app/BUILD.bazel app library target.
This is the reason why your build tests for app are failling.
![image](https://user-images.githubusercontent.com/76404817/183595408-40696eab-0cd3-4756-923b-3a1e4c711a97.png)
Thanks

vrajdesai78 committed:
InjectableBottomSheetFragment added in Bazel

yash10019coder commented on app/src/main/java/org/oppia/android/app/fragment/BUILD.bazel:
```diff
@@ -91,3 +91,18 @@ kt_android_library(
+        "//app/src/main/java/org/oppia/android/app/view:view_component",
+        "//app/src/main/java/org/oppia/android/app/view:view_component_builder_injector",
+        "//app/src/main/java/org/oppia/android/app/view:view_component_factory",
+        "//third_party:androidx_appcompat_appcompat",
```
```suggestion
        "//third_party:androidx_appcompat_appcompat",
        "//third_party:com_google_android_material_material",
```

vrajdesai78 replied:
Updated thanks

yash10019coder replied:
@rt4914 can you please resolve this comment I am not able to resolve it

yash10019coder reviewed (changes requested):
Hey @vrajdesai78 your build is failling beacuase you haven't added this dependcy for androidx dialog by google PTAL thanks

rt4914 reviewed (changes requested):
@vrajdesai78 Please have a look at @yash10019coder comment. Also will now depend on @BenHenning help for bazel.

vrajdesai78 committed:
material dependency added in bazel

vrajdesai78 committed:
Rerun ci-checks

rt4914 reviewed (approved):
LGTM, Thanks.

vrajdesai78 committed:
BottomSheet added for RevisionCard

vrajdesai78 committed:
Imports optimized

vrajdesai78 commented:
I have added code for revision card activity and updated tests screenshots and video accordingly.

rt4914 commented on app/src/main/res/drawable/ic_baseline_close_24.xml:
```diff
@@ -0,0 +1,10 @@
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24"
+    android:tint="?attr/colorControlNormal">
```
This tint needs be removed.

vrajdesai78 replied:
Done

rt4914 commented on app/src/main/res/drawable/ic_bottomsheet.xml:
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+  android:width="24dp"
+  android:height="24dp"
+  android:tint="#FFFFFF"
```
This tint can be removed.

vrajdesai78 replied:
Done

rt4914 commented on app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml:
```diff
@@ -0,0 +1,44 @@
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:orientation="vertical"
+    android:padding="10dp">
```
Padding should be in multiple of `4dp`.

vrajdesai78 replied:
Updated to 8dp

rt4914 commented on app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml:
```diff
@@ -0,0 +1,44 @@
+      android:id="@+id/action_options"
+      android:layout_width="match_parent"
+      android:layout_height="48dp"
+      android:drawablePadding="10dp"
```
Padding should be either `8dp` or `12dp`

vrajdesai78 replied:
Updated to 8dp

rt4914 commented on app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml:
```diff
@@ -0,0 +1,44 @@
+      android:drawablePadding="10dp"
+      android:gravity="center_vertical"
+      android:text="@string/menu_options"
+      android:textColor="@color/color_def_black"
```
File an issue for `Dark Mode Support for bottom sheet` in both the places and mention it in your PR

vrajdesai78 replied:
Added TODO for that and filed issue #4513

rt4914 commented on app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml:
```diff
@@ -0,0 +1,44 @@
+    <TextView
+      android:id="@+id/action_options"
+      android:layout_width="match_parent"
+      android:layout_height="48dp"
```
height should be `wrap_content` and `min_height` should be `48dp`.

vrajdesai78 replied:
Updated

rt4914 commented on app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml:
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:app="http://schemas.android.com/apk/res-auto">
+
+  <androidx.appcompat.widget.LinearLayoutCompat xmlns:android="http://schemas.android.com/apk/res/android"
```
There is some indentation issue  ( I am fully sure)

vrajdesai78 replied:
Updated with proper indentation

rt4914 commented on app/src/main/res/layout/exploration_activity.xml:
```diff
@@ -53,8 +53,18 @@
+            android:id="@+id/action_bottom_sheet_options_menu"
+            style="?android:actionOverflowButtonStyle"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
```
Does it need contentDescription for screen-readers?

vrajdesai78 replied:
By default it has content description `Button, More options` so I don't think we need to add contentDescription explicitly.

rt4914 commented on app/src/main/res/layout/revision_card_activity.xml:
```diff
@@ -35,6 +35,16 @@
+          android:id="@+id/action_bottom_sheet_options_menu"
+          style="?android:actionOverflowButtonStyle"
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
```
Does it need contentDescription for screen-readers?

rt4914 replied:
If yes, please add test too.

vrajdesai78 replied:
As mentioned earlier I don't think we need to add contentDescription explicitly.

rt4914 reviewed (changes requested):
@vrajdesai78 Suggested design related changes.

Also please generate **RTL** screenshots for proof that it works correctly.

vrajdesai78 committed:
Fixed nit changes in xml files

vrajdesai78 commented:
> @vrajdesai78 Suggested design related changes.
> 
> Also please generate **RTL** screenshots for proof that it works correctly.

Made all suggested changes and also added **RTL** screenshots in PR description.

vrajdesai78 committed:
Rerun ci checks

vrajdesai78 committed:
RTL support added for close

vrajdesai78 commented:
@rt4914 Added new screenshots for **RTL** after updating strings.xml

rt4914 commented on app/src/main/res/values-pt-rBR/strings.xml:
```diff
@@ -13,6 +13,7 @@
   <string name="menu_help">Ajuda</string>
   <string name="exploration_activity_title">Reprodutor de Exploração</string>
   <string name="help_activity_title">Ajuda</string>
+  <string name="bottom_sheet_options_menu_close">Fechar</string>
```
Please don't add strings for other languages, that's a separate process.

vrajdesai78 replied:
Removed

rt4914 commented on app/src/main/res/values-sw/strings.xml:
```diff
@@ -10,6 +10,7 @@
   <string name="menu_help">Msaada</string>
   <string name="exploration_activity_title">Kicheza Ugunduzi</string>
   <string name="help_activity_title">Msaada</string>
+  <string name="bottom_sheet_options_menu_close">Funga</string>
```
Please don't add strings for other languages, that's a separate process.

vrajdesai78 replied:
Removed

rt4914 commented on app/src/main/res/values-ar/strings.xml:
```diff
@@ -10,6 +10,7 @@
   <string name="menu_help">مساعدة</string>
   <string name="exploration_activity_title">مشغل رحلة الاستكشاف</string>
   <string name="help_activity_title">المساعدة</string>
+  <string name="bottom_sheet_options_menu_close">إغلاق</string>
```
Please don't add strings for other languages, that's a separate process.

vrajdesai78 replied:
Removed

rt4914 reviewed (approved):
LGTM, Thanks.

vrajdesai78 committed:
Revert "RTL support added for close"

This reverts commit 49c295cde32b19be028e1653102cd9207a4bcae9.

yash10019coder reviewed (approved):
LGTM @vrajdesai78 thanks

BenHenning commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -104,6 +103,11 @@ class ExplorationActivityPresenter @Inject constructor(
 
+    binding.actionBottomSheetOptionsMenu.setOnClickListener {
+      val bottomSheetOptionsMenu = BottomSheetOptionsMenu()
+      bottomSheetOptionsMenu.show(activity.supportFragmentManager, bottomSheetOptionsMenu.tag)
```
Is there a showNow that can be used, instead?

vrajdesai78 replied:
Done

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -255,6 +256,8 @@ exempted_file_path: "app/src/main/java/org/oppia/android/app/player/audio/Cellul
 exempted_file_path: "app/src/main/java/org/oppia/android/app/player/audio/CellularDataInterface.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/player/audio/LanguageDialogFragment.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/player/audio/LanguageInterface.kt"
+exempted_file_path: "app/src/main/java/org/oppia/android/app/player/exploration/BottomSheetOptionsMenu.kt"
```
Could this maybe be tested directly? I think you could test all the options for the sheet directly, then just verify that it shows/hides when expected within ExplorationActivityTest.

vrajdesai78 replied:
@BenHenning can you give any reference test in which we are testing any functionality directly like here we want to test BottomSheetOptionsMenu. I know how to write test case w.r.t. activity, but here I guess you suggesting to do in a generalized way.

BenHenning replied:
Sure. https://github.com/oppia/oppia-android/blob/0b095f19a0f4a16b9049a662f649e5c1268e0b0a/app/src/test/java/org/oppia/android/app/translation/AppLanguageWatcherMixinTest.kt#L129 is one such example, though there may be others. ``TestActivity`` is a general activity that can be used to test individual components (you could have a test helper that adds the menu directly, and then you can test UI interactions & callbacks from within that).

BenHenning replied:
This is still unresolved--we need a BottomSheetOptionsMenuTest to avoid needing this exemption.

vrajdesai78 replied:
Added `BottomSheetOptionsMenuTest` as per your suggestion

BenHenning reviewed (changes requested):
Thanks @vrajdesai78! Just had a couple of comments per my codeowners files. PTAL.

vrajdesai78 committed:
showNow updated

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityPresenter.kt:
```diff
@@ -62,6 +62,11 @@ class RevisionCardActivityPresenter @Inject constructor(
 
+    binding.actionBottomSheetOptionsMenu.setOnClickListener {
+      val bottomSheetOptionsMenu = BottomSheetOptionsMenu()
+      bottomSheetOptionsMenu.show(activity.supportFragmentManager, bottomSheetOptionsMenu.tag)
```
This should also be ``showNow``--ditto for other showings in the app.

vrajdesai78 replied:
Done

BenHenning commented on app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml:
```diff
@@ -0,0 +1,48 @@
+  xmlns:app="http://schemas.android.com/apk/res-auto">
+
+  <androidx.appcompat.widget.LinearLayoutCompat
+    android:id="@+id/bottom_sheet_layout"
```
Suggest: ``options_menu_bottonm_sheet_container`` (since IDs are global, it's best to make them really specific).

vrajdesai78 replied:
Updated

BenHenning reviewed (commented):
Thanks @vrajdesai78! Just a few comments left from me--PTAL.

vrajdesai78 commented:
Keep this PR Open

vrajdesai78 committed:
Merge branch 'develop' into Bottomsheet-for-overflow-menu

vrajdesai78 commented:
@BenHenning @rt4914 I have tried to write test case for BottomSheetOptionsMenu, but I am facing error as shown below.  [Code](https://gist.github.com/vrajdesai78/0d5ea7f2eef389c9f7e3d4739267069e)

![Screenshot from 2022-09-09 00-28-17](https://user-images.githubusercontent.com/43074241/189204279-593b8a82-01d6-4846-8e17-dea5d98e33a5.png)

BenHenning commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -101,6 +102,13 @@ class ExplorationActivityPresenter @Inject constructor(
+      bottomSheetOptionsMenu.showNow(activity.supportFragmentManager, bottomSheetOptionsMenu.tag)
+    }
+
+    updateToolbarTitle(explorationId)
```
I think this was accidentally put back after a conflict with the latest develop changes. This should be removed as it's being set below (after field initialization to avoid referencing uninitialized values).

Ditto for ``internalProfileId``--we no longer need that.

vrajdesai78 replied:
I have fixed it @BenHenning

BenHenning reviewed (commented):
@vrajdesai78 did you investigate the error? A quick look at the PR shows that you included extra lines that you shouldn't have after resolving recent merge conflicts from develop.

vrajdesai78 committed:
nit changes

vrajdesai78 committed:
failing tests fixed

rt4914 commented:
@vrajdesai78 Your CI checks are passing now. Also please reply all open comments from Ben and assign it to him for review.

vrajdesai78 committed:
Test added for BottomSheetOptionsMenu

vrajdesai78 committed:
Optimized Code

vrajdesai78 committed:
nit changes

vrajdesai78 committed:
Made suggested changes

vrajdesai78 committed:
optimized code

vrajdesai78 committed:
Fixed static checks

vrajdesai78 committed:
Fixed failing checks

vrajdesai78 commented:
@BenHenning can you help me with failing CI checks. New tests which I have written are passing locally with gradle, but it's creating issue in CI checks with Bazel.

BenHenning commented on scripts/assets/kdoc_validity_exemptions.textproto:
```diff
@@ -248,6 +248,8 @@ exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/BindableAda
 exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/BindableAdapterTestFragmentPresenter.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/BindableAdapterTestViewModel.kt"
+exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/BottomSheetOptionsMenuTestActivity.kt"
+exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/BottomSheetOptionsMenuTestActivityPresenter.kt"
```
Please add KDocs for these rather than requesting exemptions.

vrajdesai78 replied:
Done

BenHenning commented on scripts/assets/accessibility_label_exemptions.textproto:
```diff
@@ -9,6 +9,7 @@ exempted_activity: "app/src/main/java/org/oppia/android/app/player/state/testing
 exempted_activity: "app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivity"
 exempted_activity: "app/src/main/java/org/oppia/android/app/testing/AudioFragmentTestActivity"
 exempted_activity: "app/src/main/java/org/oppia/android/app/testing/BindableAdapterTestActivity"
+exempted_activity: "app/src/main/java/org/oppia/android/app/testing/BottomSheetOptionsMenuTestActivity"
```
No need to exempt this. We can use test_activity_label for the label to avoid needing the exemption.

vrajdesai78 replied:
Done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/testing/BottomSheetOptionsMenuTestActivityTest.kt:
```diff
@@ -0,0 +1,222 @@
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(application = BottomSheetOptionsMenuTestActivityTest.TestApplication::class)
+class BottomSheetOptionsMenuTestActivityTest {
```
Should this be BottomSheetOptionsMenuTest?

vrajdesai78 replied:
Created new test with `BottomSheetOptionsMenuTest`

BenHenning commented on app/src/main/res/layout/test_bottom_sheet_options_menu.xml:
```diff
@@ -0,0 +1,19 @@
+    android:minHeight="48dp"
+    android:scaleType="center" />
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Please add EOF newline.

vrajdesai78 replied:
Removed File

BenHenning commented on app/src/main/java/org/oppia/android/app/testing/BottomSheetOptionsMenuTestActivityPresenter.kt:
```diff
@@ -0,0 +1,53 @@
+    }
+  }
+
+  fun handleOnOptionsItemSelected(itemId: Int): Boolean {
```
Instead of intenting here, could we maybe introduce an interface & a mock, then call it to pass the item ID to the test? References:

https://github.com/oppia/oppia-android/blob/d22f502dbf01b21dd9b54002dd564b23f1c12043/app/src/main/java/org/oppia/android/app/notice/testing/BetaNoticeDialogFragmentTestActivity.kt#L24

https://github.com/oppia/oppia-android/blob/1fcd7746f23af5e546558203eb685d0c7662a39d/app/src/sharedTest/java/org/oppia/android/app/notice/BetaNoticeDialogFragmentTest.kt#L205

https://github.com/oppia/oppia-android/blob/1fcd7746f23af5e546558203eb685d0c7662a39d/app/src/sharedTest/java/org/oppia/android/app/notice/BetaNoticeDialogFragmentTest.kt#L173

vrajdesai78 replied:
Removed this test and added mock in new `BottomSheetMenuTestActivity`

BenHenning commented on app/src/main/res/layout/test_bottom_sheet_options_menu.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Does this need to be a constraint layout or could we just make the ImageButton top-level?

BenHenning replied:
Resolved by the file being removed.

BenHenning commented on app/src/main/res/layout/test_bottom_sheet_options_menu.xml:
```diff
@@ -0,0 +1,19 @@
+  android:layout_height="match_parent"
+  xmlns:app="http://schemas.android.com/apk/res-auto">
+
+  <ImageButton
```
Could this just be a Button or even a TextView? Does it need an image, or to be clickable?

vrajdesai78 replied:
Removed File

BenHenning commented on app/src/main/res/layout/test_bottom_sheet_options_menu.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Is this file actually being used?

vrajdesai78 replied:
Deleted

BenHenning reviewed (changes requested):
Thanks @vrajdesai78. Took another pass.

Regarding the failing CI checks, have you looked at the errors being reported by the failing tests? It's likely that you'll need to make some changes to Bazel files as part of introducing these new components. See https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions for details on getting Bazel set up locally if you don't already have it set up.

vrajdesai78 committed:
Created new BottomSheetOptionsMenuTest

vrajdesai78 committed:
Fixed ktlint issues

vrajdesai78 commented:
@BenHenning I have added new BUILD.bazel file in `player/exploratino/testing/BUILD.bazel`. I hava also added it in sharedTest folder. So, now as I have directly added `BottomSheetOptionsMenuTest` inside `player/exploration/BottomSheetOptionsMenuTest.kt` will I need to add `ExplorationActivityTest.kt` to BUILD.bazel as `ExplorationActivityTest.kt` and `BottomSheetOptionsMenuTest.kt` are in same package.

vrajdesai78 committed:
Fixed checks in bazel

vrajdesai78 committed:
Optimized code

vrajdesai78 committed:
Added dagger_rules() in bazel

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/BUILD.bazel:
```diff
@@ -0,0 +1,46 @@
+load("//app:test_with_resources.bzl", "test_with_resources")
+
+app_test(
+    name = "BottomSheetOptionsMenuTest",
```
If you're going to add this Bazel file, you'll also need to add a test target for the existing test (ExplorationActivityTest).

vrajdesai78 replied:
Done

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/BUILD.bazel:
```diff
@@ -0,0 +1,46 @@
+app_test(
+    name = "BottomSheetOptionsMenuTest",
+    processed_src = test_with_resources("BottomSheetOptionsMenutest.kt"),
+    test_class = "org.oppia.android.app.notice.BottomSheetOptionsMenuTest",
```
This should correspond to the package of BottomSheetOptionsMenuTest.

BenHenning replied:
This has been addressed per the latest code.

BenHenning reviewed (commented):
Thanks @vrajdesai78! This LGTM, just have a few nits PTAL.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/BUILD.bazel:
```diff
@@ -0,0 +1,46 @@
+
+app_test(
+    name = "BottomSheetOptionsMenuTest",
+    processed_src = test_with_resources("BottomSheetOptionsMenutest.kt"),
```
I think you have a typo in the test name (which is causing your CI failure).

vrajdesai78 replied:
Updated

vrajdesai78 committed:
Nit changes in BUILD.bazel

vrajdesai78 committed:
Merged with latest develop

vrajdesai78 committed:
Updated Bazel

vrajdesai78 committed:
Merged with latest develop

vrajdesai78 commented:
@BenHenning I have added ExplorationActivityTest in Bazel. Can you help me with failing tests, it's failing because of the following reason. 
```
unresolved reference: nav_app_bar_navigate_up_description
      onView(withContentDescription(R.string.nav_app_bar_navigate_up_description)).perform(click())
```

BenHenning commented:
> @BenHenning I have added ExplorationActivityTest in Bazel. Can you help me with failing tests, it's failing because of the following reason.
> 
> ```
> unresolved reference: nav_app_bar_navigate_up_description
>       onView(withContentDescription(R.string.nav_app_bar_navigate_up_description)).perform(click())
> ```

@vrajdesai78 it looks like nav_app_bar_navigate_up_description comes from AndroidX. One thing that can help here is comparing the deps list of your new test target to the previous deps that the test was using:

https://github.com/oppia/oppia-android/blob/7e4f2ddb6b0f94883b39ee8fe45108e206d91e52/app/BUILD.bazel#L823

One of the dependencies in TEST_DEPS needs to also be in the deps list of the new ExplorationActivityTest target to resolve the missing dependency here. I suggest looking specifically at the AndroidX ones, since a quick Google search suggested that the string is part of one of those libraries (core maybe?).

vrajdesai78 committed:
androidx core deps added

vrajdesai78 committed:
dependency added in bazel

vrajdesai78 committed:
depdencies updated in bazel

vrajdesai78 committed:
androidx dependencies added

vrajdesai78 committed:
TestRunner dependency added

vrajdesai78 committed:
databinding deps added in bazel

vrajdesai78 committed:
annotation deps added

vrajdesai78 committed:
annotation dep added to bazel

vrajdesai78 committed:
All androidx test deps added

vrajdesai78 committed:
commit to re-run checks

vrajdesai78 commented:
@BenHenning now why CI checks are failing, previously when I tried with same deps all CI checks have passed.

BenHenning commented:
@vrajdesai78 it's possible these are flakes, but we'll need to wait for the run to finish to restart the failing ones. These would be new flakes that I haven't seen before, though.

BenHenning commented:
@vrajdesai78 it looks like it was a transient failure.

BenHenning reviewed (commented):
@vrajdesai78 please make sure everything is done & replied to in this PR, then assign it back.

vrajdesai78 commented:
> @vrajdesai78 please make sure everything is done & replied to in this PR, then assign it back.

@BenHenning I have replied to all open comments.

vrajdesai78 commented:
> > @BenHenning I have added ExplorationActivityTest in Bazel. Can you help me with failing tests, it's failing because of the following reason.
> > ```
> > unresolved reference: nav_app_bar_navigate_up_description
> >       onView(withContentDescription(R.string.nav_app_bar_navigate_up_description)).perform(click())
> > ```
> 
> @vrajdesai78 it looks like nav_app_bar_navigate_up_description comes from AndroidX. One thing that can help here is comparing the deps list of your new test target to the previous deps that the test was using:
> 
> https://github.com/oppia/oppia-android/blob/7e4f2ddb6b0f94883b39ee8fe45108e206d91e52/app/BUILD.bazel#L823
> 
> One of the dependencies in TEST_DEPS needs to also be in the deps list of the new ExplorationActivityTest target to resolve the missing dependency here. I suggest looking specifically at the AndroidX ones, since a quick Google search suggested that the string is part of one of those libraries (core maybe?).

Added androidx dependencies

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/BUILD.bazel:
```diff
@@ -0,0 +1,102 @@
+"""
+Tests for notices shown in the app.
```
I thought I left a comment about this, but can't seem to find it. This is out of date (probably from a copy & paste?)--please update.

vrajdesai78 replied:
Updated, Thanks for correcting me

BenHenning reviewed (commented):
Thanks @vrajdesai78! The PR LGTM except for one doc string that needs to be updated (not sure how this was missed in my earlier passes).

vrajdesai78 committed:
Kdoc updated

BenHenning reviewed (commented):
Thanks @vrajdesai78! The PR LGTM, but CI seems broken for BottomSheetOptionsMenuTest and looking at the error, I suspect it's due to merging with the latest develop. Can you please fix the issue in the PR (such that CI is passing) and then send it back for review?

vrajdesai78 committed:
Merged with develop

vrajdesai78 commented:
> Thanks @vrajdesai78! The PR LGTM, but CI seems broken for BottomSheetOptionsMenuTest and looking at the error, I suspect it's due to merging with the latest develop. Can you please fix the issue in the PR (such that CI is passing) and then send it back for review?

Merged with latest develop and solved merge conflicts

vrajdesai78 committed:
Remove practicetab

vrajdesai78 committed:
Fixed menu icon

BenHenning commented on app/src/main/res/layout/revision_card_activity.xml:
```diff
@@ -29,19 +29,29 @@
-            android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
-            android:layout_marginEnd="20dp" />
+        <TextView
+          android:id="@+id/revision_card_toolbar_title"
+          style="@style/MarqueeToolbarTitleTextView"
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
+          android:layout_marginEnd="20dp" />
```
I think you accidentally reverted the spacing adjustment here--please fix.

vrajdesai78 replied:
Done

BenHenning reviewed (commented):
Thanks @vrajdesai78. The PR LGTM, but there was one small spacing issue that came up after the sync with develop--PTAL.

vrajdesai78 committed:
Removed extra spacing

BenHenning reviewed (approved):
Thanks @vrajdesai78! LGTM. Just waiting on CI.

BenHenning commented:
CI is passing, so going ahead and merging this.
